### PR TITLE
[consensus] Make leader reputation db fetching work if we ask for older target_rounds

### PR DESCRIPTION
**Pull request by ikabiljo**
And improving caching, to reduce number of DB calls.

For #1395

### CODE REVIEW

File: consensus/src/liveness/leader_reputation.rs [hunk 1]
```diff
@@ -25,70 +25,108 @@ pub trait MetadataBackend: Send + Sync {
...
-            .unwrap_or((0, 0));
-        if !(known_round == target_round
-            || known_version == self.aptos_db.get_latest_version().unwrap_or(0))
-        {
-            if let Err(e) = self.refresh_window(target_round) {
-                error!(
-                    error = ?e, "[leader reputation] Fail to refresh window",
-                );
-                return vec![];
+        let db_result = self.db_result.lock();
```
Comment (sitalkedia) [inline consensus/src/liveness/leader_reputation.rs:104]: This acts like a cache for fetching the result, I am wondering why this is needed as we already introduced a cache for proposers in https://github.com/aptos-labs/aptos-core/pull/1474?

File: consensus/src/liveness/leader_reputation.rs [hunk 2]
```diff
@@ -25,70 +25,108 @@ pub trait MetadataBackend: Send + Sync {
...
+        let events = &db_result.0;
+        let version = db_result.1;
+        let hit_end = db_result.2;
+
+        let has_larger = events.first().map_or(false, |e| e.round() >= target_round);
+        let lastest_db_version = self.aptos_db.get_latest_version().unwrap_or(0);
+        // check if fresher data has potential to give us different result
+        if !has_larger && version < lastest_db_version {
+            // we need to drop the lock before refreshing.
+            drop(db_result);
```
Comment (sitalkedia) [inline consensus/src/liveness/leader_reputation.rs:114]: Hmm..dropping the lock here means we might have multiple concurrent `refresh_db_result` calls - semantically that should be fine but we unnecessarily make additional DB call. Does it make sense to drop the lock in `refresh_db_result` call instead?

**Review by sitalkedia — COMMENTED** [review]
LGTM

**Review by sitalkedia — APPROVED** [review]

File: consensus/src/liveness/leader_reputation.rs [hunk 3]
```diff
@@ -25,70 +25,117 @@ pub trait MetadataBackend: Send + Sync {
...
-            .unwrap_or((0, 0));
-        if !(known_round == target_round
-            || known_version == self.aptos_db.get_latest_version().unwrap_or(0))
-        {
-            if let Err(e) = self.refresh_window(target_round) {
-                error!(
-                    error = ?e, "[leader reputation] Fail to refresh window",
-                );
-                return vec![];
+        let locked = self.db_result.lock();
```
Comment (zekun000) [inline consensus/src/liveness/leader_reputation.rs:115]: probably remove the comment on L113

**Pull request merged**